File: scalar/crt.c

#include <stdio.h>

long gcd(long m, long n) {
    long t;
    while (n != 0){
	    t = n;
	    n = m % n;
	    m = t;
    }
    return m;
}

long lcm(long m, long n) {
    return m * n / gcd(m,n);
}

int main(int argc, char* argv[]) {
    int rep = 10000;
    if(argc >=2) {
	rep = atoi (argv[1]);
    }

    int n = 40;
    long residual[n];
    long divisor[n];
    int i;
    for(i=0; i < n; i++) {
	residual[i] = i+1;
	divisor[i] = i+2;
    }
    long a,r;
    int iter;
    for(iter = 0; iter < rep; iter++){
	a=divisor[0];
	r=residual[0];
	for(i = 1; i < n; i++){
	    while(r % divisor[i] != residual[i]) {
		r = r+a;
	    }
	    a = lcm(a, divisor[i]);
	    //printf("a=%ld\n",a);
	}
    }
    printf("%ld\n",r);
    return 0;
}

File: scalar/crt.R
# TODO: Add comment
# 
# Author: Administrator
###############################################################################

# I define GCD and LCM inside the run function

setup <- function(args='100') {
    n<-as.integer(args[1])
    if(is.na(n)){ n <- 100 }
    return(n)
}

run <- function(repcount=100) {
	gcd = function(m, n) {
		while (n != 0){
			t <- n;
			n <- m %% n;
			m <- t; 
		}
		m;
	}
	
	lcm = function(m,n) {
		m * n / gcd(m,n);
	}
	
	#construct the input
	n<-40
	residual <- 1:n; #n length vec
	divisor <- residual+1; #n length vec

	
	#print(divisor);
	#print(residual);
	#benchpart
	for(iter in 1:repcount){
		a <- divisor[1];
		r <- residual[1];
		for(i in 2:n) {
			
			while(r %% divisor[i] != residual[i]) {
				r <- r + a;
				#print("a/r=");print(a);print(r);
			}
			a <- lcm(a, divisor[i]);
			#print("a=");print(a);
		}
	}
	print(r);
}

if (!exists('harness_argc')) {
    n <- setup(commandArgs(TRUE))
    run(n)
}
